compiler/src/dotty/tools/dotc/typer/TypeAssigner.scala

```diff
@@ -401,15 +401,19 @@ trait TypeAssigner {
   def assignType(tree: untpd.CaseDef, pat: Tree, body: Tree)(using Context): CaseDef = {
     val ownType =
       if (body.isType) {
-        val params = new TreeAccumulator[mutable.ListBuffer[TypeSymbol]] {
+        val getParams = new TreeAccumulator[mutable.ListBuffer[TypeSymbol]] {
           def apply(ps: mutable.ListBuffer[TypeSymbol], t: Tree)(using Context) = t match {
             case t: Bind if t.symbol.isType => foldOver(ps += t.symbol.asType, t)
             case _ => foldOver(ps, t)
           }
         }
-        HKTypeLambda.fromParams(
-          params(new mutable.ListBuffer[TypeSymbol](), pat).toList,
-          defn.MatchCase(pat.tpe, body.tpe))
+        val params = getParams(new mutable.ListBuffer[TypeSymbol](), pat).toList
+        pat.tpe match
+          case AppliedType(tycon, args) =>
+            val tparams = tycon.typeParamSymbols
+            for param <- params do param.info = param.info.subst(tparams, args)
```

Comment:
I don't know what this does exactly but you almost certainly shouldn't actually mutate types and instead should create a copy of the params with different types.

Reply:
Gladly.  How do I do that?

Reply:
param.copy(info = ...)

Reply:
oh, but those are pattern-bound symbols? Then I'm not sure if they should be copied or not. If they could be created with the correct type that'd be much better.

Reply:
I wish I could create them with the correct type, but I don't think I can, in an order independent way.  I iterate over them with their expected type, and then I need to substitute any references among those pattern-bound symbol types.  Unless I'm missing something?

Reply:
If you create them and then mutate their info in the same block of code, that would be fine.